Constrained split

DaveShuckerow opened the pull request:
This allows us to get the size of the split children and prevent layout overflows at small screen sizes.

kenzieschmoll commented on packages/devtools_app/lib/src/flutter/split.dart:
```diff
@@ -46,11 +71,37 @@ class Split extends StatefulWidget {
+  /// In other words, `(firstBuilder == null) != (firstChild == null)`.
   final Widget firstChild;
 
   /// The child that will be laid out last along [axis].
```
nit: r/last/second to match builder comment

DaveShuckerow replied:
Done, thanks for catching this!

kenzieschmoll commented on packages/devtools_app/test/flutter/split_test.dart:
```diff
@@ -288,3 +344,15 @@ Split buildSplit(Axis axis, {@required double initialFirstFraction}) {
 }
+
+Split buildSplitWithBuilder(Axis axis,
+    {@required double initialFirstFraction}) {
```
nit: trailing comma

DaveShuckerow replied:
Done.

kenzieschmoll commented on packages/devtools_app/test/flutter/split_test.dart:
```diff
@@ -54,6 +54,20 @@ void main() {
+        expect(tester.element(find.byKey(_w1)).size, const Size(0, 600));
+        expect(tester.element(find.byKey(_w2)).size, const Size(790, 600));
+        expect(tester.element(find.byKey(split.dividerKey)).size,
+            const Size(10, 600));
```
use `equals()` matcher here and elsewhere

DaveShuckerow replied:
Is there any value that `equals` to the meaning of the code?

When given a value for the second parameter, `expect` will use an `equals` matcher by default.

Using it explicitly would make most of these expect statements run to two lines without changing the meaning of the code.  I think that the file will be more readable as-is.

DaveShuckerow replied:
Is there any value that `equals` to the meaning of the code?

When given a value for the second parameter, `expect` will use an `equals` matcher by default.

Using it explicitly would make most of these expect statements run to two lines without changing the meaning of the code.  I think that the file will be more readable as-is.

jacob314 commented:
Can't you achieve the same effect by wrapping each of the child widgets in a
`LayoutBuilder` or am I missing something?
If that is the case it might be cleaner to define a simple ConstrainedSplit
ConstraintedSplit stateless widget that composes together the existing Split widget along with a couple of LayoutBuilder and a couple of SizedBuilder callbacks to build each subtree.

DaveShuckerow commented:
> Can't you achieve the same effect by wrapping each of the child widgets in a
> `LayoutBuilder` or am I missing something?
> If that is the case it might be cleaner to define a simple ConstrainedSplit
> ConstraintedSplit stateless widget that composes together the existing Split widget along with a couple of LayoutBuilder and a couple of SizedBuilder callbacks to build each subtree.

The trick is that the children of the split need to know how large they are to avoid overflow at small sizes.  Without touching the split at all, it's possible to wrap its children in LayoutBuilders and get the same behavior:

```dart
Split(
  firstChild: LayoutBuilder(builder: (context, constraints)) {
    final size = Size(constraints.maxWidth, constraints.maxHeight);
    // do stuff with size to avoid an overflow
  }
),

Split(
  firstBuilder: (context, size) {
    // do stuff with size to avoid an overflow
  },
),
``` 

However, LayoutBuilders are less efficient than just passing the size down if we already know what the size of each child is supposed to be.  LayoutBuilders build as a second pass of build calls, but with this approach we only make one pass.

jacob314 commented:
> > Can't you achieve the same effect by wrapping each of the child widgets in a
> > `LayoutBuilder` or am I missing something?
> > If that is the case it might be cleaner to define a simple ConstrainedSplit
> > ConstraintedSplit stateless widget that composes together the existing Split widget along with a couple of LayoutBuilder and a couple of SizedBuilder callbacks to build each subtree.
> 
> The trick is that the children of the split need to know how large they are to avoid overflow at small sizes. Without touching the split at all, it's possible to wrap its children in LayoutBuilders and get the same behavior:
> 
> ```dart
> Split(
>   firstChild: LayoutBuilder(builder: (context, constraints)) {
>     final size = Size(constraints.maxWidth, constraints.maxHeight);
>     // do stuff with size to avoid an overflow
>   }
> ),
> 
> Split(
>   firstBuilder: (context, size) {
>     // do stuff with size to avoid an overflow
>   },
> ),
> ```
> 
> However, LayoutBuilders are less efficient than just passing the size down if we already know what the size of each child is supposed to be. LayoutBuilders build as a second pass of build calls, but with this approach we only make one pass.

I agree this will be slightly less efficient but I'm a bit skeptical it will be sufficiently less efficient for that to matter. For example, I expect we will likely have a maximum of around 5 splitters on a page. For example, if there were cases where we want a splitter for every row in a table then I would be convinced. Otherwise, it would seem more consistent with Flutter to favor composition over adding extra functionality to a single widget.

DaveShuckerow commented:
> I agree this will be slightly less efficient but I'm a bit skeptical it will be sufficiently less efficient for that to matter. 

That's fair.  We can put this PR on ice until we decide it's time for a round of performance improvements.

@kenzieschmoll you can get the same API that this was going to add with the following:

```dart
Split(
  firstChild: LayoutBuilder(builder: (context, constraints)) {
    final size = Size(constraints.maxWidth, constraints.maxHeight);
    // do stuff with size to avoid an overflow
  }
),

Split(
  firstBuilder: (context, size) {
    // do stuff with size to avoid an overflow
  },
),
```

jacob314 commented:
Feel free to add a helper widget that wraps both children with a LayoutBuilder to achieve the same effect without the performance improvement.

DaveShuckerow commented:
> Feel free to add a helper widget that wraps both children with a LayoutBuilder to achieve the same effect without the performance improvement.

That doesn't help.  The child content has to know how big it is to avoid creating an overflow.

For example,
```dart
Split(
  secondChild: LayoutBuilder(builder: (context, constraints) {
    return SizedBox(width: 300, height: 400, child: Text('content'));
  }),
),
```

This does nothing.  The layout builder is only useful if the constraints it passes down are used by the child content to size itself:

```dart
Split(
  secondChild: LayoutBuilder(builder: (context, constraints) {
    final desiredSize = Size(300, 400);
    final child = SizedBox.fromSize(size: desiredSize, child: Text('content'));
    if (constraints.maxHeight < desiredSize.height) {
      return SingleChildScrollView(child: child);
    }
    return Align(alignment: Alignment.bottomCenter, child: child);
  }),
),
```

jacob314 commented:
Sorry I was unclear.
What i mean is you could define a `ConstrainedSplit` widget as follows using composition to layer in the useful constrained split behavior. As a bonus, if performance was an issue you could later change the underlying implementation to avoid the extra LayoutBuilder widgets.

```dart
class ConstrainedSplit extends StatelessWidget {
   final ConstrainedBuilder builder1;
   final ConstrainedBuilder builder2;
   ConstraintedSplit(this.builder1, this.builder2, ... other split args);

   Widget build(BuildContext context) {
     return Split(
       other split args,
       firstChild: LayoutBuilder(builder: builder1)
       secondChild: LayoutBuilder(builder: builder2),
     );
  }
}
```
